### ISSUE TITLE
Kafka Streams Transaction ID expiration

### ISSUE THREAD

**Issue by matthewi** [Apicurio/apicurio-registry#1134]
Kafka streams has a default timeout of 7 days for a streams transaction ID. It seems if your schemas are not seeing a lot of churn that Kafka expires the transaction ID which causes an error in the registry service:

```
2020-12-29 08:24:37,992 ERROR [org.apa.kaf.str.pro.int.StreamThread] (schemaregistry-9f2e3bbb-9756-40bc-8729-9f01701bee43-StreamThread-2) stream-thread [schemaregistry-9f2e3bbb-9756-40bc-8729-9f01701bee43-StreamThread-2] Encountered the following unexpected Kafka exception during processing, this usually indicate Streams internal errors:: org.apache.kafka.streams.errors.StreamsException: task [1_5] Abort sending since an error caught with a previous record (timestamp 1609230277933) to topic global-id-store-changelog due to org.apache.kafka.common.errors.InvalidPidMappingException: The producer attempted to use a producer id which is not currently assigned to its transactional id.
        at org.apache.kafka.streams.processor.internals.RecordCollectorImpl.recordSendError(RecordCollectorImpl.java:154)
        at org.apache.kafka.streams.processor.internals.RecordCollectorImpl.access$500(RecordCollectorImpl.java:52)
        at org.apache.kafka.streams.processor.internals.RecordCollectorImpl$1.onCompletion(RecordCollectorImpl.java:214)
        at org.apache.kafka.clients.producer.KafkaProducer$InterceptorCallback.onCompletion(KafkaProducer.java:1356)
        at org.apache.kafka.clients.producer.internals.ProducerBatch.completeFutureAndFireCallbacks(ProducerBatch.java:231)
        at org.apache.kafka.clients.producer.internals.ProducerBatch.abort(ProducerBatch.java:159)
        at org.apache.kafka.clients.producer.internals.RecordAccumulator.abortUndrainedBatches(RecordAccumulator.java:799)
        at org.apache.kafka.clients.producer.internals.Sender.maybeSendAndPollTransactionalRequest(Sender.java:424)
        at org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:312)
        at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:239)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.kafka.common.errors.InvalidPidMappingException: The producer attempted to use a producer id which is not currently assigned to its trans
actional id.
```

This spins for a while before the health check fails:
```
2020-12-29 08:28:08,073 INFO  [io.api.reg.met.PersistenceExceptionLivenessCheck] (executor-thread-120) After this event, the error counter is 1 (out of the maximum 5 allowed).
2020-12-29 08:28:08,069 WARN  [io.api.reg.met.PersistenceExceptionLivenessCheck] (executor-thread-120) Liveness problem suspected in PersistenceExceptionLivenessCheck because of an exception: : java.lang.IllegalStateException: KafkaStreams is not running. State is ERROR.
```

Minimally this seems like something the registry server should handle more gracefully as its not unexpected. Alternatively,  a keepalive that held the transaction id open should avoid this situation. A mature project in production is very likely to have very low schema churn.

Unfortunately a lot of registry clients don't hand this gracefully and this can trigger a cascading service failure.

**Comment by EricWittmann** [comment]
Thoughts on this, @alesj ?

**Comment by alesj** [comment]
Hmmm ... it would be good if we could (somehow) reproduce this?
e.g. set the tx timeout to 5min (or something even less)

@tombentley suggested (among other things) there might be a hang somewhere in our Streams processing ... but I don't see it ...
(WaitForDataService, afais, should not block processing ...)

**Comment by matthewi** [comment]
Some additionally background/clarification: In this case we're using the ccompat api since we're dealing with avro schemas for kafka topics. We're on AKS and using strimzi 0.19 and kafka 2.5.

**Comment by matthewi** [comment]
I put together a test scenario and I'm able to reproduce it consistently. Based on logs there's some question of if the schema publish that triggers the error condition actually gets persisted. I suspect not but have not confirmed.

**Comment by alesj** [comment]
> I put together a test scenario and I'm able to reproduce it consistently

Can you share this, so I can try to reproduce it locally as well.

**Comment by matthewi** [comment]
So, I tried to package this up with our kafka configuration in case there was an issue there. I wasn't able to reproduce it using docker-compose -- only running under kubernetes. In docker it seems the service restarts the kafka client like one would expect when it encounters this error. I'm thinking under k8s the health check kicks in and kills the service before it has a chance to restart the kafka client. This triggers a lengthy reconciliation of the topic which leads to potentially lengthy unavailability.

**Comment by matthewi** [comment]
For reference, here are the relevant kafka configuration properties:
```
      default.replication.factor: 3
      log.cleanup.policy: compact
      log.message.format.version: 2.3.1
      num.partitions: 8
      offsets.topic.replication.factor: 3
      transaction.state.log.min.isr: 2
      transaction.state.log.replication.factor: 3
      transactional.id.expiration.ms: 5000
```

We're running Kafka 2.5 via strimzi on k8s. 

I'll link a gist with the files for setting environment on k8s.

**Comment by matthewi** [comment]
https://gist.github.com/matthewi/8025a87faf47c5e8d95d585159957a92

**Comment by matthewi** [comment]
Logs:

https://gist.github.com/matthewi/69a5e3fefc7b44c7374f7de96475d1bc

**Comment by matthewi** [comment]
Does this by chance only fall apart if you try to replicate the streams topics? Our default topic configuration is 8 partitions+3 replicas. I've not been able to trigger this yet if I switch to the 1 partition 1 replica that's in the docs

**Comment by EricWittmann** [comment]
@tombentley @alesj Does this new information help diagnose this at all?

**Comment by matthewi** [comment]
I was able to get it to trigger on the single partition last night; I missed it since the recovery is <1 minute and doesn't always trigger a pod restart. I'm concerned about resiliency in this configuration if we lose a broker, but at this point that's a less likely occurrence than hitting this issue so I'm going to run with that config for now.

IF there is other information I can provide or tests you would like me to run, please lmk.

I was also wondering if the ksql work would potentially impact this?

**Comment by EricWittmann** [comment]
If you were able to test the `kafkasql` storage found on the `master` branch that would very interesting.  The Kafka SQL storage is probably as simple a usage of Kafka as possible for this use-case (since I don't have the Kafka expertise that Ales has).  It's just a very simple consumer/producer rather than a Streams app.  It only requires a single Kafka topic and it supports log compaction to constrain the size of the log as much as possible.  There are tradeoffs between the two approaches though.  The Kafka SQL implementation probably scales a little bit better for reads, and Search will be faster.  However, it does replicate the full state on **every** node, which the Streams implementation doesn't do (it spreads out the information across the cluster).  Also the Kafka SQL implementation will take longer to start up a new node the more messages are in the log, since it must replay the entire log to rebuild the in-memory state.

So yeah... tradeoffs.  :)

**Comment by matthewi** [comment]
Interesting. Can you quantify long? The shortest time I've observed with Streams is about 1 minute, but sometimes as much as 25. 

On my test instance it does seem to take quite a while on restart with ~1000 schemas. This might not be too bad if I could run multiple instances, but the health checks don't seem to be accurately reflecting the state of the service:

```
ksql $ curl http://localhost:8080/health/live

{
    "status": "UP",
    "checks": [
        {
            "name": "PersistenceTimeoutReadinessCheck",
            "status": "UP",
            "data": {
                "errorCount": 0
            }
        },
        {
            "name": "StorageLivenessCheck",
            "status": "UP"
        },
        {
            "name": "PersistenceExceptionLivenessCheck",
            "status": "UP",
            "data": {
                "errorCount": 0
            }
        },
        {
            "name": "ResponseErrorLivenessCheck",
            "status": "UP",
            "data": {
                "errorCount": 0
            }
        }
    ]
}
ksql $ curl http://localhost:8080/health/ready

{
    "status": "UP",
    "checks": [
        {
            "name": "PersistenceSimpleReadinessCheck",
            "status": "UP"
        },
        {
            "name": "Database connections health check",
            "status": "UP"
        },
        {
            "name": "ResponseTimeoutReadinessCheck",
            "status": "UP",
            "data": {
                "errorCount": 0
            }
        }
    ]
}
```

**Comment by matthewi** [comment]
5 minutes 🤔 Does that scale linearly with the number of schemas?
```
2021-02-18 17:58:03,808 CONFIG [com.sun.net.httpserver] (main) HttpServer created http null
2021-02-18 18:02:38,318 DEBUG [org.apa.kaf.cli.con.int.ConsumerCoordinator] (KSQL Kafka Consumer Thread) [Consumer clientId=consumer-78c45793-23dd-4ab8-9857-f53b1eed492e-1, groupId=78c45793-23dd-4ab8-9857-f53b1eed492e] Completed asynchronous auto-commit of offsets {kafkasql-journal-0=OffsetAndMetadata{offset=1009, leaderEpoch=0, metadata=''}}
```

**Comment by EricWittmann** [comment]
It will probably scale linearly with the number of schemas, yes.  Or rather it will scale with the number of *changes* to the registry.  We do add tombstone messages when we can (e.g. when deleting an artifact).  But that helps only so much.  Five minutes for 1000 schemas is slower than what I've seen in my local testing (less than a minute for 10k schemas) - but maybe I am testing on a faster machine?

You are right about the health/liveness checks - there is a TODO in the KafkaSql implementation to properly implement the "isReady" check to only be `true` once the initial load period is complete.  I don't think that should be too hard to implement, but it got put on the back burner in favor of other priorities.  Good opportunity for a contribution??  =) 

https://github.com/Apicurio/apicurio-registry/blob/master/storage/kafkasql/src/main/java/io/apicurio/registry/storage/impl/kafkasql/KafkaSqlRegistryStorage.java#L247

**Comment by matthewi** [comment]
Thanks Eric. Running 2 instances of kafkasql has the startup and recovery time down to 1-2 minutes for 2K schemas (I programatically created a a bunch just to have some data to play with). Hopefully this isn't an artifact of docker not scrubbing the disk when pods restart.

**Comment by EricWittmann** [comment]
When I get a moment I'll try again and see what sort of numbers I get.  There are some gatling scripts checked into the repository that can be used for this purpose.

I've been thinking about the readiness check and how to implement that.  I was thinking about publishing a message to the topic on startup.  Then have an "isReady" flag (initialized to `false` ) that turns `true` only when that startup message is consumed.  There would probably still be the potential for a little bit of a race condition, since other nodes could publish additional messages during the 1-2 minutes of startup time.  But it's probably acceptable based on expected usage (relatively few writes).

@famartinrh Did you have any other ideas around implementing kafkasql readiness?

**Comment by famartinrh** [comment]
Something similar to your idea but instead of using a special message using the kafka records timestamp, not reporting isReady=true until a record with a more recent timestamp than the server startup timestamp is consumed. With this approach we will need some way to handle cases where there is no records in the topic or no new records are produced, but still I think it will be better than producing a special message.

**Comment by EricWittmann** [comment]
Also it assumes that the kafka record timestamp is correct relative to the time/clock of the consumer.  Doing this sort of operation based on timestamps is always tricky for this reason.  Usually system clocks are fairly close (only off by a few seconds) but I worry about timezone issues.

If we use a special message we can always tombstone that message to clean it up after.  And as you point out, we don't need to have special handling for "no messages consumed".

**Comment by famartinrh** [comment]
I still think there should be a better way of doing this. Kafka is known for allowing to replay events, it should have some API that allows us to do what we need to do now...

I need to investigate it further but the KafkaConsumer API has methods such as `commited` or `endOffsets` that will return the current latest offset for each partition. We could take that information at the beginning and make the consumer thread report if those offsets have been read.

**Comment by EricWittmann** [comment]
That sounds promising.  I'm very open to other approaches, especially Kafka specific ones!

**Comment by matthewi** [comment]
I was speculating a bit on leveraging the checkpoint mechanism, but I'm
liking the direction this is going.

**Comment by EricWittmann** [comment]
FYI I am implementing my dumb bootstrap-message approach now.  That was a 30 minute task and we can easily swap it out later for something better.  But I also wanted to report back here on some startup times.  I used a gatling script I created awhile ago to load up the registry with a bunch of artifacts.  Startup times listed below for different numbers of artifacts:

* 10,000 artifacts -> 9s
* 20,000 artifacts -> 12s
* 30,000 artifacts -> 17s
* 40,000 artifacts -> 18s
* 100,000 artifacts -> 46s

This is a small sample size, and obviously your mileage may vary depending on the speed of the node.  I tested this locally on this:

![image](https://user-images.githubusercontent.com/1890703/109703321-5308da80-7b63-11eb-8b5e-e1fbe678164d.png)

I was running Kafka and Registry on the same machine.

**Comment by EricWittmann** [comment]
I'm closing this as we have released 2.0.0 and 2.0.1 now without the Kafka Streams storage implementation.  Because we're no longer using Streams, my understanding is that we aren't using transactions and therefore this issue is not relevant.  If I'm wrong we can re-open this!
